block/base/mempool.go

```diff
@@ -39,11 +39,25 @@ type (
 		// txCache is a map of all transactions in the mempool. It is used
 		// to quickly check if a transaction is already in the mempool.
 		txCache map[string]struct{}
+
+		// comparator is the comparison function used to verify the ordering of txs in ProcessProposal
+		comparator Comparator
 	}
+
+	// Comparator is a comparison function that the mempool uses to verify the ordering of txs. That is,
+	// if this > other (return 1, nil), then this will be ordered before other. Conversely, if this < other,
+	// this will be ordered after other.
+	Comparator func(ctx sdk.Context, this, other sdk.Tx) (int, error)
 )
```

Comment:
Although I do think this is maybe the simplest fix, I think that it introduces a new interface that will probably end up confusing developers as there are now three different comparator functions:
1. txpriority
2. the mempool sequence tie logic + txpriority
3. the new comparator function.

Alternatively, we could just implement a new processlane handler that doesnt verify the order of txs (slightly prefer this). Since we anticipate a future where we will likely have to design our own mempool im fine with this change as things stand today.

Reply:
Yeah to me the TxPriority is the way that we've already defined to tell the mempool implementation how to order the transactions. So I think it makes the most sense to just pass in a TxPriority which returns 0 or equal for all tx comparisons. It's functionally the same as what you have here afaict, but doesn't introduce a new interface.